decoder.go

```diff
@@ -0,0 +1,211 @@
+package redis
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+// Decoder represents a type with a custom Decode() function that can
+// take a string and parse it to set a desired value on a struct field,
+// similar to how encoding/json implements UnmarshalJSON().
+type Decoder interface {
+	Decode(string) error
+}
+
+// defDecoder represents decoding functions for default built-in types.
+type defDecoder func(reflect.Value, string) error
+
+// structField represents a single field in a target struct.
+type structField struct {
+	index int
+
+	// Decoder.Decode() or defDecoder()
+	fn interface{}
+}
+
+var (
+	defDecoders = map[reflect.Kind]defDecoder{
+		reflect.Bool:    decodeBool,
+		reflect.Int:     decodeInt,
+		reflect.Int8:    decodeInt,
+		reflect.Int16:   decodeInt,
+		reflect.Int32:   decodeInt,
+		reflect.Int64:   decodeInt,
+		reflect.Uint:    decodeUint,
+		reflect.Uint8:   decodeUint,
+		reflect.Uint16:  decodeUint,
+		reflect.Uint32:  decodeUint,
+		reflect.Uint64:  decodeUint,
+		reflect.Float32: decodeFloat,
+		reflect.Float64: decodeFloat,
+		reflect.Slice:   decodeStringSlice,
+		reflect.String:  decodeString,
+	}
+
+	// Global map of struct field specs that is populated once for every new
+	// struct type. This caches the field types and the corresponding decoder
+	// functions to avoid iterating through struct fields on every scan.
+	// {type: { structFieldTag:  }}
+	structSpecs = make(map[reflect.Type]map[string]*structField)
+	mut         sync.RWMutex
+)
+
+// Scan scans the results from a key-value Redis map result set ([]interface{})
+// to a destination struct. The Redis keys are matched to the struct's field
+// tags defined in the struct indifcated by fieldTag.
+// eg: fieldTag=redis where Name string `redis:"name"`
+func Scan(vals []interface{}, dest interface{}, fieldTag string) error {
```

Comment:
Do we need to support custom `fieldTag` here? I would drop it and use `redis`, because otherwise you need a separate struct cache for each different `fieldTag`.